src/graphnet/models/model.py

```diff
@@ -39,6 +39,7 @@ def _construct_trainers(
         log_every_n_steps: int = 1,
         gradient_clip_val: Optional[float] = None,
         distribution_strategy: Optional[str] = "ddp",
+        inference_strategy: Optional[str] = "auto",
```

Comment:
I'm not a fan of introducing this argument.  Also, as it it's written now, the argument is not passed to the method - effectively hardcoding it to `"auto"`. 

I think a better solution would be a small refactor, where `_construct_trainers` is changed to

```
def _construct_trainer(
        self,
        max_epochs: int = 10,
        gpus: Optional[Union[List[int], int]] = None,
        callbacks: Optional[List[Callback]] = None,
        ckpt_path: Optional[str] = None,
        logger: Optional[LightningLogger] = None,
        log_every_n_steps: int = 1,
        gradient_clip_val: Optional[float] = None,
        distribution_strategy: Optional[str] = "ddp",
        **trainer_kwargs: Any,
    ) -> None:

        if gpus:
            accelerator = "gpu"
            devices = gpus
        else:
            accelerator = "cpu"
            devices = 1

        trainer = Trainer(
            accelerator=accelerator,
            devices=devices,
            max_epochs=max_epochs,
            callbacks=callbacks,
            log_every_n_steps=log_every_n_steps,
            logger=logger,
            gradient_clip_val=gradient_clip_val,
            strategy=distribution_strategy,
            ckpt_path = ckpt_path,
            **trainer_kwargs,
        )
   return trainer
```

This way, the function returns a single trainer based on a single set of `Trainer` arguments. We can then call this whenever the user interacts with this functionality through `Model`. So in `model.fit` we adjust:

```
def fit(
        self,
        train_dataloader: DataLoader,
        val_dataloader: Optional[DataLoader] = None,
        *,
        max_epochs: int = 10,
        gpus: Optional[Union[List[int], int]] = None,
        callbacks: Optional[List[Callback]] = None,
        ckpt_path: Optional[str] = None,
        logger: Optional[LightningLogger] = None,
        log_every_n_steps: int = 1,
        gradient_clip_val: Optional[float] = None,
        distribution_strategy: Optional[str] = "ddp",
        **trainer_kwargs: Any,
    ) -> None:
        """Fit `Model` using `pytorch_lightning.Trainer`."""
        # Checks
        if callbacks is None:
            callbacks = self._create_default_callbacks(
                val_dataloader=val_dataloader,
            )
        elif val_dataloader is not None:
            callbacks = self._add_early_stopping(
                val_dataloader=val_dataloader, callbacks=callbacks
            )

        self.train(mode=True)
        trainer = self._construct_trainer(
            max_epochs=max_epochs,
            gpus=gpus,
            callbacks=callbacks,
            ckpt_path=ckpt_path,
            logger=logger,
            log_every_n_steps=log_every_n_steps,
            gradient_clip_val=gradient_clip_val,
            distribution_strategy=distribution_strategy,
            **trainer_kwargs,
        )

        try:
            trainer.fit(
                self, train_dataloader, val_dataloader, ckpt_path=ckpt_path
            )
        except KeyboardInterrupt:
            self.warning("[ctrl+c] Exiting gracefully.")
            pass

```

and in `Model.predict`:

 ```
def predict(
        self,
        dataloader: DataLoader,
        gpus: Optional[Union[List[int], int]] = None,
        distribution_strategy: Optional[str] = "auto",
    ) -> List[Tensor]:
        """Return predictions for `dataloader`.

        Returns a list of Tensors, one for each model output.
        """
        self.train(mode=False)

        callbacks = self._create_default_callbacks(
                val_dataloader=None,
            )
        
        inference_trainer = self._construct_trainer(gpus=gpus, 
                                                    distribution_strategy=distribution_strategy, 
                                                    callbacks = callbacks)

        predictions_list = inference_trainer.predict(self, dataloader)
        assert len(predictions_list), "Got no predictions"

        nb_outputs = len(predictions_list[0])
        predictions: List[Tensor] = [
            torch.cat([preds[ix] for preds in predictions_list], dim=0)
            for ix in range(nb_outputs)
        ]

        return predictions
```

This _should_ also mean that there no references to `DataLoader` in `Model`. Hopefully this removes your pickle error.

Reply:
Thank you for your feedback. I was in doubt as to how we would want to implement a workaround for the lack of an inference strategy. I like the refactoring, and I will get back and see if this resolves the Pickle bug.